src/Compilers/CSharp/Portable/Syntax/SyntaxExtensions.cs

```diff
@@ -398,84 +401,116 @@ internal static bool IsOutVarDeclaration(this DeclarationExpressionSyntax p)
                         break;
+                    case SyntaxKind.GenericName:
+                        var genericNameSyntax = (GenericNameSyntax)type;
+                        for (int i = genericNameSyntax.TypeArgumentList.Arguments.Count - 1; i >= 0; i--)
+                        {
+                            stack.Push(new ArrayRankSpecifierOrTypeSyntax(genericNameSyntax.TypeArgumentList.Arguments[i]));
+                        }
+                        break;
+                    case SyntaxKind.QualifiedName:
+                        var qualifiedNameSyntax = (QualifiedNameSyntax)type;
+                        stack.Push(new ArrayRankSpecifierOrTypeSyntax(qualifiedNameSyntax.Right));
+                        stack.Push(new ArrayRankSpecifierOrTypeSyntax(qualifiedNameSyntax.Left));
+                        break;
+                    case SyntaxKind.AliasQualifiedName:
+                        var aliasQualifiedNameSyntax = (AliasQualifiedNameSyntax)type;
+                        stack.Push(new ArrayRankSpecifierOrTypeSyntax(aliasQualifiedNameSyntax.Name));
+                        break;
+                    case SyntaxKind.IdentifierName:
+                    case SyntaxKind.OmittedTypeArgument:
+                    case SyntaxKind.PredefinedType:
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(type.Kind());
+                }
+            }
 
-                    for (int index = 0; index < argsCount - 1; index++)
-                    {
-                        var typeArgument = genericNameSyntax.TypeArgumentList.Arguments[index];
-                        typeArgument.VisitRankSpecifiers(action, argument);
-                    }
-
-                    type = genericNameSyntax.TypeArgumentList.Arguments[argsCount - 1];
-                    goto recurse;
-                case SyntaxKind.QualifiedName:
-                    var qualifiedNameSyntax = (QualifiedNameSyntax)type;
-                    qualifiedNameSyntax.Left.VisitRankSpecifiers(action, argument);
-                    type = qualifiedNameSyntax.Right;
-                    goto recurse;
-                case SyntaxKind.AliasQualifiedName:
-                    var aliasQualifiedNameSyntax = (AliasQualifiedNameSyntax)type;
-                    type = aliasQualifiedNameSyntax.Name;
-                    goto recurse;
-                case SyntaxKind.IdentifierName:
-                case SyntaxKind.OmittedTypeArgument:
-                case SyntaxKind.PredefinedType:
-                    break;
-                default:
-                    throw ExceptionUtilities.UnexpectedValue(type.Kind());
+            stack.Free();
+        }
+
+        private struct ArrayRankSpecifierOrTypeSyntax
+        {
+            private readonly ArrayRankSpecifierSyntax? ArrayRank;
+            private readonly TypeSyntax? Type;
```

Comment:
Could we track a single `SyntaxNode` instead? Then the `struct` could be removed.